### PR TITLE
Adding ReinforceGamma

### PR DESCRIPTION
**Pull request by ssampang**
I was thinking it would be nice to test the gradient calculation more thoroughly - possibly by calculating it numerically and then comparing, but I couldn't figure out how to deal with the random nature of the function.

### CODE REVIEW

**Comment by ssampang** [comment]
Also not sure why git is stating that I've written all of test.lua myself

**Comment by GOATmessi8** [comment]
@ssampang  I wonder why the return of updataGradInput is "self.gradInput", since all your computation are "gradShape" and "gradScale"?

**Comment by ssampang** [comment]
@ruinmessi you can see the logic for that [here](https://github.com/Element-Research/dpnn/pull/51/commits/03a6c63311c81beb7029e800630863307ac13c3b#diff-dbf12896bf902ed91d3c8c2c692926c0R73). Essentially, if the input is a tensor, then we assume its the shape, and we use the scale variable set at initialization. If the input is a table, then we assume the user is providing both shape and scale, and so we must return a self.gradInput that is a table where the first element is a tensor of the gradient with respect to shape, and the second element is a tensor of the gradient with respect to the scale. On lines 73-77, we set local variables equal to the relevant fields of self.gradInput, and so when we modify gradShape (and gradScale if a scale was provided as input), we are modifying self.gradInput as well.

**Comment by nicholas-leonard** [comment]
@ssampang The real test.lua is in test/test.lua. Yours is another file. Could your rebase or submit a new PR?

**Comment by nicholas-leonard** [comment]
@ssampang As for a valid unit test for these kind of modules, you could add a test that makes sure it learns what it is supposed to. I think you mentioned this before. As in https://github.com/Element-Research/dpnn/blob/master/test/test.lua#L1578-L1691, you can put the test under `dpnnbigtest.ReinforceGamma` which you can test using `th -ldpnn -e "dpnn.bigtest{'ReinforceGamma'}".

File: ReinforceGamma.lua [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+   if torch.type(scale) == 'number' then
+      scale = shape.new():resizeAs(shape):fill(scale)
+   else
+      if not scale:dim() == shape:dim() then
+         scale:copy(self.__scale)
+      end
+   end
+   gradShape:copy(cephes.digamma(shape:float()))
+   gradShape:mul(-1)
+   gradShape:add( torch.log(self.output) )
```
Comment (nicholas-leonard) [inline ReinforceGamma.lua:95]: @ssampang Make sure to use buffers here. So instead of `torch.log(self.output)`, you could have something like : 

``` lua
self._buffer = self._buffer or self.output.new() -- lazy memory allocation
self._buffer:log(self.output)
```

And don't forget to nil the buffer in type:

``` lua
function ReinforceGamma:type(type, cache)
    self._buffer = nil
    return parent.type(self, type, cache)
end
```

**Comment by ssampang** [comment]
@nicholas-leonard could you explain the motivation for the buffer?

**Comment by nicholas-leonard** [comment]
Allocates memory once per object instead of every call to function. Doesn't
rely on garbage collection to manage memory.
